documentation/smg/17_using_pixel_data.md

```diff
@@ -0,0 +1,394 @@
+>> sub_pix.num_pixels
+    ans =
+      1000
+```
+
+These methods provide ways to retrieve data from the full pixel data array,
+whether that data is in memory or is in a file.
+They also have the advantage of being able to get/set multiple columns at once.
+This can be more efficient, as you're performing a single slice, instead of
+multiple.
+
+```matlab
+% Below examples are equivalent, but using the get_data method is more
+% efficient as it performs fewer slices
+
+% Performs one slice over the pixel array
+>> run_sig_var = pix.get_data({'run_idx', 'signal', 'variance'});
+
+% Performs one slice for each property access
+>> run_sig_var = zeros(3, pix.num_pixels);
+>> run_sig_var(1, :) = pix.run_idx;
+>> run_sig_var(2, :) = pix.signal;
+>> run_sig_var(3, :) = pix.variance;
+```
+
+All getter/setter methods use [_absolute indexing_](#Absolute-and-Relative-Indexing).
+
+#### Absolute and Relative Indexing
+
+Absolute indexing refers to indexing into the full pixel array.
+Relative indexing refers to indexing into the current page of pixels.
+
+So, suppose a `PixelData` object interfaces a file which contains 100 pixels,
+with a page size of 10 pixels.
+Then the following holds:
+
+```matlab
+>> pix.num_pixels
+     ans =
+       100
+>> pix.move_to_page(3);
+>> all(pix.signal(1:6) == pix.get_data('signal', 21:26))
+     ans =
+       true
+```
+
+There exists a private utility function that can perform the conversion between
+these index types.
+
+See
+[ADR 17](../adr/0017-separate-absolute-and-relative-indexing-APIs-in-pixel-array.md)
+for the decision record on this.
+
+## Implementing File-Backed Algorithms
+
+### Simple Example
+
+Suppose we wish to subtract some constant background from an SQW object.
+This requires that we subtract some constant from the pixels' signal.
+For this we could implement a `minus` function,
+(in reality we have a general `do_binary_op` function).
+This can be performed in a `while` loop over pixels.
+
+```matlab
+classdef PixelData < handle
+
+    methods
+
+        ...
+
+        function minus(obj, scalar)
+            %MINUS Subtract a scalar from every pixel's signal
+            %
+            % Input:
+            % -----
+            % scalar  A scalar double.
+            %
+
+            % Never assume we're on the first page.
+            % If we are already on the first page, this call only costs one
+            % integer comparison
+            obj.move_to_first_page();
+
+            % Subtract scalar from every value in cache.
+            % If there is no data in the cache, this call will load that data
+            % from file and then perform the operation.
+            obj.signal = obj.signal - scalar;
+
+            % Check if there are any more pixels to load from file.
+            while obj.has_more()
+                % Move to the next page of data.
+                % This call dumps the previous page to a temporary file and
+                % marks the page as "dirty".
```

Comment:
Does it always mark it as dirty, even if no changes have taken place? Is it possible to set the dirty marking to any setters/getters?

Reply:
No, pages are only marked dirty if they're edited.

Reply:
Not entirely sure what you mean by the second question

Reply:
It looked like you were saying dirty was set on calling `advance`. If you set "dirty" on anything that modifies pixels (presumably the pixels' setter) then that would avoid that.